### PR TITLE
NodalVoidVolume added to Geochemistry

### PR DESCRIPTION
**Pull request by WilkAndy**
A bit more for @loganharbour .   This is more complicated than the previous PR because it involves MPI stuff.  Fortunately i could mostly copy-and-paste-and-modify from something similar in PorousFlow.

### CODE REVIEW

**Comment by loganharbour** [comment]
Ahh something I'm more familiar with! I'll get to this on Monday.

**Comment by WilkAndy** [comment]
Thanks @loganharbour .    It's actually quite fun code, i think

**Comment by WilkAndy** [comment]
Hi @loganharbour .  If you can do this PR before the other (reactive transport) PR, that'd be good.  This is more of a core utility and a much smaller PR.

a

**Comment by WilkAndy** [comment]
Not stale - waiting for a review.

**Comment by WilkAndy** [comment]
I'm back from holidays, so it'd be good if you could review this, @loganharbour

**Comment by loganharbour** [comment]
On it right now! We've even got testing back so we can merge again!

**Review by loganharbour — CHANGES_REQUESTED** [review]
This all looks great but it seems like you can replace your complex structure with just:
```
std::unordered_map<const Node *, Real> _nodal_void_volume;
```

**Comment by loganharbour** [comment]
I'm going to hold off on specific comments until the above is addressed because it would change the code a bit.

**Comment by WilkAndy** [comment]
> This all looks great but it seems like you can replace your complex structure with just:
> 
> ```
> std::unordered_map<const Node *, Real> _nodal_void_volume;
> ```

Yes, i tried that with a similar object in porous flow, but it was disappointingly slow.  Hence all the screwing around with "my node number" stuff, which admittedly uses memory, but is a lot faster.   So i'd like to keep the current version unless you really object.

**Comment by loganharbour** [comment]
> > This all looks great but it seems like you can replace your complex structure with just:
> > ```
> > std::unordered_map<const Node *, Real> _nodal_void_volume;
> > ```
> 
> Yes, i tried that with a similar object in porous flow, but it was disappointingly slow. 

Really? I don't understand how the performance should be any different. Both require a map lookup.

**Comment by WilkAndy** [comment]
> > > This all looks great but it seems like you can replace your complex structure with just:
> > > ```
> > > std::unordered_map<const Node *, Real> _nodal_void_volume;
> > > ```
> > 
> > 
> > Yes, i tried that with a similar object in porous flow, but it was disappointingly slow.
> 
> Really? I don't understand how the performance should be any different. Both require a map lookup.

Ooops, i remembered the code incorrectly.   Sure, i'll make the changes, if it's possible (the _nodes_to_receive stuff has to be re-written too)

**Comment by loganharbour** [comment]
I would definitely prefer that if possible! It makes it much easier to understand without the additional indexing.

**Comment by WilkAndy** [comment]
There you go, @loganharbour .   I hope this is what you  were thinking of!

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (loganharbour) [inline modules/geochemistry/src/userobjects/NodalVoidVolume.C:128]:
```suggestion
  std::map<processor_id_type, std::set<const Node *>> filled_nodes;
  std::map<processor_id_type, std::vector<dof_id_type>> global_node_nums_to_receive;
  _nodes_to_receive.clear();
  // run through all elements known by this processor (the owned + 1 layer of ghosted elements),
  // recording nodes that are attached to elements that aren't owned by this processor
  for (const auto & elem : _fe_problem.getEvaluableElementRange())
    if (this->hasBlocks(elem->subdomain_id()))
    {
      const processor_id_type elem_pid = elem->processor_id();
      if (elem_pid != _my_pid)
      {
        auto & pid_nodes = filled_nodes[elem_pid];
        for (const auto & node : elem->node_ref_range())
          if (pid_nodes.insert(&node).second) // true if we haven't filled this node
          {
            global_node_nums_to_receive[elem_pid].push_back(node.id());
            _nodes_to_receive[elem_pid].push_back(&node);
          }
      }
    }
```
Adds a set to do the checking if you've filled a node already instead of doing the linear search on `global_node_nums_to_receive` This one's not as important as the other map suggested changes, but I think it cleans it up a little!

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (loganharbour) [inline modules/geochemistry/src/userobjects/NodalVoidVolume.C:146]:
```suggestion
    auto & pid_entry = _nodes_to_send[pid];
    pid_entry.reserve(kv.second.size());
    for (const auto & node_num : kv.second)
      pid_entry.push_back(_mesh.nodePtr(node_num));
```
Here you lookup in the map multiple times - this removes that. Also reserve space ahead of time in the vector.

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (loganharbour) [inline modules/geochemistry/src/userobjects/NodalVoidVolume.C:160]:
```suggestion
    auto & pid_entry = nvv_to_send[pid];
    pid_entry.reserve(kv.second.size());
    for (const auto & nd : kv.second)
      pid_entry.push_back(_nodal_void_volume.at(nd));
```
Same here. Avoids multiple map lookups and reserves.

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (loganharbour) [inline modules/geochemistry/src/userobjects/NodalVoidVolume.C:171]:
```suggestion
    auto & receive_pid_entry = _nodes_to_receive[pid];
    const std::size_t msg_size = nvv_received.size();
    mooseAssert(msg_size == receive_pid_entry.size(),
                "Message size, " << msg_size
                                 << ", incompatible with nodes_to_receive, which has size "
                                 << receive_pid_entry.size());
    for (std::size_t i = 0; i < msg_size; ++i)
      _nodal_void_volume[receive_pid_entry[i]] += nvv_received[i];
```
Avoids multiple map lookups in `_nodes_to_receive` and `_nodal_void_volume`. Corrects type on the for loop.

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (loganharbour) [inline modules/geochemistry/src/userobjects/NodalVoidVolume.C:225]:
```suggestion
  auto find = _nodal_void_volume.find(node);
  if (find != _nodal_void_volume.end())
    return find->second;
  mooseError("nodal id ",
             node->id(),
             " not in NodalVoidVolume's data structures.  Perhaps the execute_on parameter of "
             "NodalVoidVolume needs to be set differently");
```
Avoids two map lookups. `find->second` is the entry in the map (if found).

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (loganharbour) [inline modules/geochemistry/src/userobjects/NodalVoidVolume.C:214]:
```suggestion
  for (const auto & node : _current_elem->node_ref_range())
  {
    auto & node_entry = _nodal_void_volume[&node];
    for (unsigned qp = 0; qp < _qrule->n_points(); ++qp)
      entry += _JxW[qp] * _coord[qp] * _phi[i][qp] * _porosity[qp];
  }
```
Here you're looking into the map a staggering `2 * _qrule->n_points()` times! This makes that only once.

**Review by loganharbour — CHANGES_REQUESTED** [review]
Mostly optimization changes for map access. You're gonna be a pro at maps after this! :-)

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (loganharbour) [inline modules/geochemistry/src/userobjects/NodalVoidVolume.C:49]:
```suggestion
    _my_pid(processor_id()),
```
One last thing. You're initializing these with their default constructors already, so you can just remove them.

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:51]: Yea, i don't want to always provide the `concentration` parameter - hopefully no one will ever have to supply it.  So i'll meet you half way on this one :-)

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:49]: OK

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:92]: Thanks

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:128]: Thanks

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:146]: Thanks

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:160]: Thanks

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:171]: Thanks

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:214]: Thanks - good optimisation here!

File: modules/geochemistry/src/userobjects/NodalVoidVolume.C
Comment (WilkAndy) [reply modules/geochemistry/src/userobjects/NodalVoidVolume.C:225]: That's a good optimisation since this method may be called quite often

**Comment by loganharbour** [comment]
Looks great to me! Will approve and merge tonight when it turns green.

**Comment by WilkAndy** [comment]
> Looks great to me! Will approve and merge tonight when it turns green.

Cool, thanks for the review @loganharbour

**Review by loganharbour — APPROVED** [review]

**Pull request merged**
